### ISSUE TITLE
[Bug] riskScore vs riskLevel mismatch — Taiwan 89/LOW, Brazil 99/CRITICAL

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#23]
## Problem
The `/api/countries` endpoint returns contradictory risk assessments. Taiwan scores 89/100 but is labeled "LOW". Brazil scores 99 and is labeled "CRITICAL" when it should be ELEVATED at most. 7 of 8 countries are CRITICAL.

Root cause: `riskScore` and `riskLevel` come from two different XGBoost outputs that can disagree:
- `riskLevel` = argmax class prediction (XGBoost picks the most likely class)
- `riskScore` = probability of that class × 100

When the model is confident in LOW but assigns high probability mass across multiple classes, you get a high score with a LOW label. And when training data was oversampled for HIGH/CRITICAL (Sprint 2 class balancing), the model over-predicts CRITICAL for countries that aren't warzones.

**Current output (broken):**
```json
{"countryCode": "TW", "country": "Taiwan", "riskScore": 89, "riskLevel": "LOW"}
{"countryCode": "BR", "country": "Brazil", "riskScore": 99, "riskLevel": "CRITICAL"}
{"countryCode": "RS", "country": "Serbia", "riskScore": 74, "riskLevel": "CRITICAL"}
```

A judge seeing Taiwan at 89/100 labeled LOW will think the system is broken. A judge seeing Brazil at 99 CRITICAL will think the model is useless.

## Goal
Make riskLevel always derived from riskScore thresholds so score and label never contradict. Fix the riskScore calculation to use a weighted probability approach that produces meaningful differentiation across countries.

### In scope
- `backend/ml/risk_scorer.py` — fix `predict_risk()` function
- New threshold-based riskLevel derivation
- New riskScore calculation that uses all class probabilities (not just argmax)
- Verify all 8 countries produce sensible, differentiated scores

### Out of scope
- XGBoost retraining (that's #22)
- Frontend changes
- Other endpoints

## Acceptance Criteria
- [ ] riskLevel is ALWAYS consistent with riskScore (no 89/LOW, no 30/CRITICAL)
- [ ] Thresholds: 0-20 LOW, 21-40 MODERATE, 41-60 ELEVATED, 61-80 HIGH, 81-100 CRITICAL
- [ ] riskScore uses weighted probability across all classes, not just argmax confidence
- [ ] Ukraine scores 85+ CRITICAL (active warzone — correct)
- [ ] Taiwan scores 40-70 range (elevated tensions, not a warzone)
- [ ] Brazil scores 20-50 range (protests/crime, not a warzone)
- [ ] Serbia scores 30-60 range (regional tensions, not critical)
- [ ] Countries show meaningful differentiation (not all 99 CRITICAL)
- [ ] /api/countries returns updated, consistent scores
- [ ] /api/risk-score returns updated scores
- [ ] /api/analyze mlMetadata uses the fixed scores

### Current predict_risk() — the broken version:
```python
def predict_risk(features: dict) -> dict:
    # ... loads model, runs prediction ...
    probabilities = model.predict_proba(X)[0]
    predicted_class = model.predict(X)[0]                    # ← argmax class
    risk_level = le.inverse_transform([predicted_class])[0]  # ← from argmax
    risk_score = int(probabilities[predicted_class] * 100)   # ← confidence in that class
    return {
        "risk_level": risk_level,      # Can be LOW
        "risk_score": risk_score,      # Can be 89
        # ← These two can DISAGREE
    }
```

The problem is `risk_score = int(probabilities[predicted_class] * 100)`. If XGBoost predicts LOW with 89% confidence, you get score=89 level=LOW. That's mathematically valid but semantically broken.

### Fixed predict_risk() — weighted probability score:

```python
RISK_LABELS = ['LOW', 'MODERATE', 'ELEVATED', 'HIGH', 'CRITICAL']

# Weights for each class — higher classes contribute more to the score
CLASS_WEIGHTS = {
    'LOW': 0,
    'MODERATE': 25,
    'ELEVATED': 50,
    'HIGH': 75,
    'CRITICAL': 100,
}

def score_from_probabilities(probabilities: list, labels: list) -> int:
    """
    Compute a 0-100 risk score as the weighted average of class probabilities.
    
    If model says: 60% LOW, 20% MODERATE, 10% ELEVATED, 5% HIGH, 5% CRITICAL
    Score = 0.60*0 + 0.20*25 + 0.10*50 + 0.05*75 + 0.05*100
         = 0 + 5 + 5 + 3.75 + 5 = 18.75 → 19 (LOW)
    
    If model says: 5% LOW, 5% MODERATE, 10% ELEVATED, 30% HIGH, 50% CRITICAL  
    Score = 0.05*0 + 0.05*25 + 0.10*50 + 0.30*75 + 0.50*100
         = 0 + 1.25 + 5 + 22.5 + 50 = 78.75 → 79 (HIGH)
    
    This always produces a score that's semantically aligned with the class distribution.
    """
    score = 0.0
    for prob, label in zip(probabilities, labels):
        score += prob * CLASS_WEIGHTS[label]
    return max(0, min(100, int(round(score))))


def level_from_score(score: int) -> str:
    """Deterministic mapping — score and level ALWAYS agree."""
    if score >= 81:
        return 'CRITICAL'
    elif score >= 61:
        return 'HIGH'
    elif score >= 41:
        return 'ELEVATED'
    elif score >= 21:
        return 'MODERATE'
    else:
        return 'LOW'


def predict_risk(features: dict) -> dict:
    model = joblib.load("models/risk_scorer.pkl")
    le = joblib.load("models/risk_label_encoder.pkl")
    scaler = joblib.load("models/feature_scaler.pkl")
    
    X = pd.DataFrame([{col: features.get(col, 0) for col in FEATURE_COLUMNS}])
    X_scaled = scaler.transform(X)
    
    probabilities = model.predict_proba(X_scaled)[0]
    labels = le.inverse_transform(range(len(probabilities)))
    
    # Weighted score from all class probabilities
    risk_score = score_from_probabilities(probabilities, labels)
    
    # Level derived from score — ALWAYS consistent
    risk_level = level_from_score(risk_score)
    
    # Confidence = probability of the derived level's class
    level_idx = list(labels).index(risk_level) if risk_level in labels else 0
    confidence = float(probabilities[level_idx])
    
    # Top feature drivers
    importance = model.feature_importances_
    top_features = sorted(zip(FEATURE_COLUMNS, importance), key=lambda x: -x[1])[:5]
    
    return {
        "risk_level": risk_level,
        "risk_score": risk_score,
        "confidence": round(confidence, 3),
        "probabilities": {
            label: round(float(prob), 3)
            for label, prob in zip(labels, probabilities)
        },
        "top_drivers": [f for f, _ in top_features],
    }
```

### Why weighted probability works:

| Scenario | Class Probabilities | Weighted Score | Level |
|----------|-------------------|----------------|-------|
| Active warzone (Ukraine) | 2% LOW, 1% MOD, 2% ELEV, 5% HIGH, 90% CRIT | 0+0.25+1+3.75+90 = 95 | CRITICAL |
| Elevated tensions (Taiwan) | 40% LOW, 15% MOD, 20% ELEV, 15% HIGH, 10% CRIT | 0+3.75+10+11.25+10 = 35 | MODERATE |
| Moderate instability (Brazil) | 30% LOW, 25% MOD, 25% ELEV, 15% HIGH, 5% CRIT | 0+6.25+12.5+11.25+5 = 35 | MODERATE |
| Peaceful country (Canada) | 90% LOW, 5% MOD, 3% ELEV, 1% HIGH, 1% CRIT | 0+1.25+1.5+0.75+1 = 4.5 | LOW |

The weighted approach captures the *distribution* of risk, not just the argmax. A country where the model is 40% LOW and 30% HIGH gets a mid-range score, not a LOW-with-high-number contradiction.

### Expected output after fix:
```json
[
  {"countryCode": "UA", "country": "Ukraine", "riskScore": 95, "riskLevel": "CRITICAL"},
  {"countryCode": "TW", "country": "Taiwan", "riskScore": 35, "riskLevel": "MODERATE"},
  {"countryCode": "IR", "country": "Iran", "riskScore": 88, "riskLevel": "CRITICAL"},
  {"countryCode": "VE", "country": "Venezuela", "riskScore": 82, "riskLevel": "CRITICAL"},
  {"countryCode": "PK", "country": "Pakistan", "riskScore": 85, "riskLevel": "CRITICAL"},
  {"countryCode": "ET", "country": "Ethiopia", "riskScore": 83, "riskLevel": "CRITICAL"},
  {"countryCode": "RS", "country": "Serbia", "riskScore": 45, "riskLevel": "ELEVATED"},
  {"countryCode": "BR", "country": "Brazil", "riskScore": 38, "riskLevel": "MODERATE"}
]
```
(Exact values will vary based on actual model probabilities — the key is differentiation and consistency.)

### Edge cases:
- If XGBoost outputs are heavily skewed (e.g., 99% in one class), weighted score will naturally be near that class's weight. Correct behavior.
- Confidence for the derived level may be low if the score falls near a threshold boundary (e.g., score=80, level=HIGH, but model gave 40% to CRITICAL). That's fine — low confidence near boundaries is honest.
- The anomaly boost in main.py (`risk_score += anomaly_score * 15`) still works — it adds to the score, and the level re-derives from the boosted score. But we need to make sure main.py re-derives the level AFTER the boost, not before.

### main.py fix needed too:
In main.py `/api/analyze`, after anomaly boost:
```python
# Step 5: Inject anomaly score and re-derive level
features['anomaly_score'] = anomaly['anomaly_score']
if anomaly['is_anomaly']:
    risk_prediction['risk_score'] = min(100, risk_prediction['risk_score'] + int(anomaly['anomaly_score'] * 15))
    # RE-DERIVE level from boosted score
    risk_prediction['risk_level'] = level_from_score(risk_prediction['risk_score'])
```

Also update `/api/countries` to use the fixed scoring.

### Hackathon tradeoff:
These exact numbers will shift again when we retrain with 195 countries (#22). But this fix ensures the fundamental contract holds: score and level always agree. The retraining will just produce better underlying probabilities.

## Manual Test Steps
1. Start backend: `uvicorn backend.main:app --reload --port 8000`
2. `curl http://localhost:8000/api/countries` — verify:
   - Ukraine: 80+ CRITICAL (or HIGH)
   - Taiwan: NOT labeled LOW if score > 20
   - Brazil: NOT labeled CRITICAL if score < 81
   - All entries: riskLevel matches riskScore thresholds
   - At least 3 different riskLevel values across 8 countries
3. `curl -X POST http://localhost:8000/api/risk-score -H "Content-Type: application/json" -d '{"country":"Taiwan","countryCode":"TW"}'` — verify riskLevel matches riskScore
4. `curl -X POST http://localhost:8000/api/risk-score -H "Content-Type: application/json" -d '{"country":"Brazil","countryCode":"BR"}'` — verify NOT 99 CRITICAL

---

## Cursor Agent Instructions

Repo: Nathan7108/HackathonPractice
Working directory: C:\Users\natha\HackathonPractice
Branch: feature/fix-risk-score-level-mismatch

### STEP 0 — Read context
```
cat PROGRESS.md
cat backend/ml/risk_scorer.py
cat backend/main.py
```

### STEP 1 — Create branch
```
git checkout main && git pull origin main
git checkout -b feature/fix-risk-score-level-mismatch
```

### STEP 2 — Fix backend/ml/risk_scorer.py
1. Add `CLASS_WEIGHTS` dict mapping each label to 0/25/50/75/100
2. Add `score_from_probabilities(probabilities, labels)` function
3. Add `level_from_score(score)` function  
4. Rewrite `predict_risk()`:
   - Get probability distribution from model
   - Compute weighted score via `score_from_probabilities()`
   - Derive level via `level_from_score()`
   - Confidence = probability of the derived level's class
5. Export `level_from_score` so main.py can import it

### STEP 3 — Fix backend/main.py
1. Import `level_from_score` from risk_scorer
2. In `/api/analyze`: after anomaly score boost, re-derive riskLevel:
   ```python
   risk_prediction['risk_level'] = level_from_score(risk_prediction['risk_score'])
   ```
3. In `/api/countries`: ensure it uses the updated predict_risk() output

### STEP 4 — Validate
```
cd C:\Users\natha\HackathonPractice
uvicorn backend.main:app --reload --port 8000
```
Then test:
```
curl http://localhost:8000/api/countries
curl -X POST http://localhost:8000/api/risk-score -H "Content-Type: application/json" -d "{\"country\":\"Taiwan\",\"countryCode\":\"TW\"}"
curl -X POST http://localhost:8000/api/risk-score -H "Content-Type: application/json" -d "{\"country\":\"Brazil\",\"countryCode\":\"BR\"}"
```
Verify: no score/level mismatches. Countries show differentiation.

### STEP 5 — Commit and push
```
git add backend/ml/risk_scorer.py backend/main.py
git commit -m "fix: derive riskLevel from riskScore thresholds — eliminate score/level mismatch"
git push -u origin feature/fix-risk-score-level-mismatch
```

### STEP 6 — Open PR
```
gh pr create --title "[Bug] Fix riskScore vs riskLevel mismatch — weighted probability scoring" --body "Closes #23" --base main
```

### Definition of Done
- [ ] No country has score > 20 with level LOW
- [ ] No country has score < 81 with level CRITICAL
- [ ] Taiwan is NOT labeled LOW
- [ ] Brazil is NOT 99 CRITICAL
- [ ] At least 3 different riskLevel values across 8 countries
- [ ] level_from_score() exported and used in main.py anomaly boost
- [ ] All 6 endpoints still work
- [ ] PR created

### If You Get Stuck
- le.inverse_transform(range(len(probabilities))) gives ordered labels matching probabilities array
- If label order doesn't match RISK_LABELS order, sort by le.transform() to align
- CLASS_WEIGHTS must use exact strings from RISK_LABELS
- The scaler (models/feature_scaler.pkl) must still be applied before predict_proba
